File: Leetcode/C++/56. Merge Intervals/solution.cpp
// https://leetcode.com/problems/merge-intervals/

#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    vector<vector<int>> merge(vector<vector<int>> &intervals)
    {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> result;
        for (auto &vec : intervals)
        {
            if (!result.empty() && vec[0] <= result.back()[1])
            {
                result.back()[1] = max(result.back()[1], vec[1]);
            }
            else
            {
                result.push_back(vec);
            }
        }
        return result;
    }
};

File: Leetcode/Typescript/56. Merge Intervals/solution.ts
// https://leetcode.com/problems/merge-intervals/description/

function merge(intervals: number[][]): number[][] {
  intervals.sort((a, b) => {
    return a[0] - b[0];
  });
  const stack: number[][] = [];

  for (const [start, end] of intervals) {
    if (stack.length === 0 || stack[stack.length - 1][1] < start) {
      stack.push([start, end]);
    } else {
      stack[stack.length - 1][1] = Math.max(stack[stack.length - 1][1], end);
    }
  }

  return stack;
}
